### DISCUSSION TITLE
Page.RunModal(0, ...) can be implemented through the "Page Management" codeunit

### DISCUSSION THREAD

**Discussion by rvanbekkum** [StefanMaron/BusinessCentral.LinterCop#258] [New Rule]
Idea based on this GitHub issue: https://github.com/DavidFeldhoff/al-codeactions/issues/30
Similar to: https://github.com/StefanMaron/BusinessCentral.LinterCop/discussions/205

The idea of the rule is to make people aware they could be using the "Page Management" codeunit instead of invoking `Page.RunModal(0, ...)` directly.

The benefits:
* Automatically an extra check for `IsGuiAllowed()`
* Automatically converts the 'Rec-Related' `Variant` to a `RecordRef`
* OnAfterGetPageID event publisher for extensions (and other similar event publishers)

You can also use this codeunit if you have a `Sales Header` record and you want it to determine the appropriate page to run (Sales Order vs. Sales Quote vs. ...).
This _could_ also be included in a rule.

**Comment by Arthurvdv** [comment]
I'm in the progress of creating this rule and would like to have some feedback over the implementation details.

A) A valid Use Case could be to run a Page without a record attached (maybe like some global variables on the page and a codeunit). The "Page Management" codeunit always expect a record, so the rule shouldn't go off when running a page without a record provided?

B) How-to handle when the page itself is set as a variable? Only let the rule go off when the PageVariable has a .SetRecord, .SetTableView ore .SetSelectionFilter before the PageVariable.Run(Model) is called?

```
    local procedure RunPage()
    var
        CustomerRec: Record Customer;
        CustomerCard: Page "Customer Card";
    begin
        // Do nothing
        Page.Run(Page::"Customer Card");

        // Raise a LC00xx rule
        Page.Run(Page::"Customer Card", CustomerRec);

        // Raise a LC00xx rule 
        Page.Run(Page::"Customer Card", CustomerRec, CustomerRec.FieldNo(Name));

        // Do nothing
        CustomerCard.Run();

        // Raise a LC00xx rule 
        CustomerCard.SetRecord(CustomerRec);
        CustomerCard.Run();
    end;
```

If there are other scenario's I'm missing here, or if you have other tips/feedback for me, I'm all ears :-)

**Reply by rvanbekkum** [reply]
A) I think `Page.Run(0)` will fail when you try to run it. You'll have to pass a record for it to determine which page to use, so only then the rule should be run.
B) I think the rule should not be run at all if the page is declared by a variable. I think the rule should only be run when we are dealing with `Page.Run(0, SomeRecord)` or `Page.RunModal(0, SomeRecord)`, because then you are saying: please just use the lookup page for this record's table. If you are declaring a page variable, you are deliberately using a specific page, so I think then the rule should not be doing anything.

**Comment by Arthurvdv** [comment]
Dit some more testing an I'm doubting what to do with a scenario as below.

```AL
    local procedure SelectCustomer()
    var
        CustomerRec: Record Customer;
    begin
        if Page.RunModal(Page::"Customer Lookup", CustomerRec) = Action::LookupOK then
            DoStuffWithSelectedCustomer(CustomerRec);
    end;
```

It seems that the "Page Management" codeunit doesn't support RunModal in combination with a Action return?

**Reply by rvanbekkum** [reply]
Sure looks like it. Good find! Might be an exception for this rule now then?

**Comment by Arthurvdv** [comment]
Pull-request created: https://github.com/StefanMaron/BusinessCentral.LinterCop/pull/275.
_(i've included some extra scenario's to exclude, see the pull-request for more details)_
